drivers/sensor/icm42688/icm42688_rtio.c

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (c) 2023 Google LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/drivers/sensor.h>
+#include "icm42688.h"
+#include "icm42688_decoder.h"
+#include "icm42688_reg.h"
+#include "icm42688_spi.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(ICM42688_RTIO, CONFIG_SENSOR_LOG_LEVEL);
+
+static inline int compute_num_samples(const enum sensor_channel *channels, size_t num_channels)
+{
+	int num_samples = 0;
+
+	for (size_t i = 0; i < num_channels; ++i) {
+		num_samples += SENSOR_CHANNEL_3_AXIS(channels[i]) ? 3 : 1;
+	}
+
+	return num_samples;
+}
+
+static inline uint32_t compute_min_buf_len(size_t num_channels, int num_output_samples)
+{
+	return sizeof(struct sensor_data_generic_header) +
+	       (num_channels * sizeof(struct sensor_data_generic_channel)) +
+	       (num_output_samples * sizeof(icm42688_sample_t));
+}
+
+static int icm42688_rtio_sample_fetch(const struct device *dev, int16_t readings[7])
+{
+	uint8_t status;
+	// struct icm42688_sensor_data *data = dev->data;
+	const struct icm42688_sensor_config *cfg = dev->config;
+	uint8_t buffer[14];
+
+	int res = icm42688_spi_read(&cfg->dev_cfg.spi, REG_INT_STATUS, &status, 1);
+
+	if (res) {
+		return res;
+	}
+
+	if (!FIELD_GET(BIT_INT_STATUS_DATA_RDY, status)) {
+		return -EBUSY;
+	}
+
+	res = icm42688_read_all(dev, buffer);
+
+	if (res) {
+		return res;
+	}
+
+	for (int i = 0; i < 7; i++) {
+		readings[i] = sys_le16_to_cpu((buffer[i * 2] << 8) | buffer[i * 2 + 1]);
+	}
+
+	return 0;
+}
+
+int icm42688_submit(const struct device *dev, struct rtio_iodev_sqe *iodev_sqe)
+{
+	const struct sensor_read_config *cfg = iodev_sqe->sqe->iodev->data;
+	const enum sensor_channel *const channels = cfg->channels;
+	const size_t num_channels = cfg->num_channels;
+	int num_output_samples = compute_num_samples(channels, num_channels);
+	uint32_t min_buf_len = compute_min_buf_len(num_channels, num_output_samples);
+	int rc;
+	uint8_t *buf;
+	uint32_t buf_len;
+	int16_t readings[7];
+
+	/* Get the buffer for the frame, it may be allocated dynamically by the rtio context */
+	rc = rtio_sqe_rx_buf(iodev_sqe, min_buf_len, min_buf_len, &buf, &buf_len);
+	if (rc != 0) {
+		LOG_ERR("Failed to get a read buffer of size %u bytes", min_buf_len);
+		rtio_iodev_sqe_err(iodev_sqe, rc);
+		return rc;
+	}
+
+	rc = icm42688_rtio_sample_fetch(dev, readings);
+	/* Check that the fetch succeeded */
+	if (rc != 0) {
+		LOG_ERR("Failed to fetch samples");
+		rtio_iodev_sqe_err(iodev_sqe, rc);
+		return rc;
+	}
+
+	rc = icm42688_encode(dev, iodev_sqe, buf, buf_len, readings);
```

Comment:
It might be faster to just read everything (all 7 readings) then just encode in the header which values were requested. Since only 7 channels exist this can be done in 7 bits. It'll waste a bit of memory, but I think it'll be much faster and simpler.